### PR TITLE
User/freerb/lv

### PR DESCRIPTION
**Pull request by BlakeFreer**

### CODE REVIEW

File: firmware/projects/LVController/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Low Voltage Controller
+
+Refer to macformula/lvcontroller
+
+the following regex can be used to extract non-empty regions of USER code in cubemx generated files.
+
+```regex
+/\* USER CODE BEGIN ([\w]*) \*/\n[\S\s]*?[\w]+[\S\s]*?/\* USER CODE END \1 \*/
```
Comment (samparent97) [inline firmware/projects/LVController/README.md:8]: Curious what the use might be here given we dont push the generated code to source control?

File: firmware/projects/LVController/inc/app.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+class Fan : public Subsystem {
+public:
+    Fan(shared::periph::DigitalOutput& enable_output,
+        shared::periph::PWMOutput& pwm_output,
+        shared::util::Mapper<float>& power_to_duty)
+        : Subsystem(enable_output),
+          pwm_output_(pwm_output),
+          power_to_duty_(power_to_duty) {}
+
+    void Enable() {
```
Comment (samparent97) [inline firmware/projects/LVController/inc/app.h:66]:
```suggestion
    void Enable() override {
```
similar for other overridden methods

File: firmware/projects/LVController/inc/app.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+     * @warning THIS IS DANGEROUS as large jumps can caused significant current
+     * rushes. Prefer SetTargetPower() and Update()
+     *
+     * @param power
+     */
+    void Dangerous_SetPowerNow(float power) {
+        SetPower(power);
+    }
+
+    void Update(float interval_sec) const {
```
Comment (samparent97) [inline firmware/projects/LVController/inc/app.h:92]: Add comment to explain how this should be used.

File: firmware/projects/LVController/inc/app.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        float current_duty = pwm_output_.GetDutyCycle();
+        float duty_step = shared::util::Clamper<float>::Evaluate(
+            target_duty_ - current_duty, -max_duty_step, +max_duty_step);
+
+        pwm_output_.SetDutyCycle(current_duty + duty_step);
+    }
+
+    bool IsAtTarget() {
+        float error = std::abs(pwm_output_.GetDutyCycle() - target_duty_);
+        return error < 0.1f;
```
Comment (samparent97) [inline firmware/projects/LVController/inc/app.h:104]: make 0.1f a constexpr

File: firmware/projects/LVController/inc/app.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+class DCDC : public Subsystem {
+public:
+    DCDC(shared::periph::DigitalOutput& enable_output,
+         shared::periph::DigitalInput& valid_input,
+         shared::periph::DigitalOutput& led_output)
+        : Subsystem(enable_output),
+          valid_input_(valid_input),
+          led_(led_output){};
+
+    bool IsValid() {
```
Comment (samparent97) [inline firmware/projects/LVController/inc/app.h:127]: Given this performs an action, prefer something like `CheckValid`

File: firmware/projects/LVController/platforms/stm32f767/bindings.cc [hunk 6]
```diff
@@ -0,0 +1,96 @@
+/// @author Blake Freer
+/// @date 2023-12-25
+
+// cubemx files
+// fw includes
+
+#include <chrono>
```
Comment (samparent97) [inline firmware/projects/LVController/platforms/stm32f767/bindings.cc:7]: Remove

File: firmware/shared/util/mappers/identity.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * @brief The most basic map; that which returns its input. This is added for
+ * completeness when a mapper object is required but has no behaviour.
+ *
+ * @tparam T Output Type
+ * @tparam U Input Type, default = `T`
+ */
+template <typename T, typename U = T>
+class IdentityMap : public shared::util::Mapper<T, U> {
```
Comment (samparent97) [inline firmware/shared/util/mappers/identity.h:18]: nice

File: firmware/projects/LVController/README.md (hunk 1 above)
Comment (BlakeFreer) [reply firmware/projects/LVController/README.md:8]: This was for analyzing last years code, will be removed

**Comment by samparent97** [comment]
Is this still a draft?

**Review by samparent97 — APPROVED** [review]

**Pull request merged**